### PR TITLE
Refactoring TOTP service

### PR DESCRIPTION
**Pull request by lashinijay**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -142,7 +155,7 @@
         <carbon.commons.version>4.8.7</carbon.commons.version>
         <carbon.commons.imp.pkg.version>[4.4.0, 5.0.0)</carbon.commons.imp.pkg.version>
         <!--Carbon identity version-->
-        <carbon.identity.framework.version>5.25.90</carbon.identity.framework.version>
+        <carbon.identity.framework.version>5.25.208</carbon.identity.framework.version>
```
Comment (lashinijay) [inline pom.xml:158]: framework version needs to be updated after [https://github.com/wso2/carbon-identity-framework/pull/4674/files](https://github.com/wso2/carbon-identity-framework/pull/4674/files) is merged.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 2]
```diff
@@ -1129,4 +1130,27 @@ public static Map<String, String> getClaimProperties(String tenantDomain, String
...
+     *
+     * @param claimURI     Claim URI.
+     * @param claimValue   Claim value.
+     * @param tenantDomain Tenant domain.
+     * @return Encrypted or plain claim value based on the EnableEncryption claim property.
+     * @throws TOTPAuthenticatorException When errors occurs while encrypting the claim value.
+     */
+    public static String getProcessedClaimValue(String claimURI, String claimValue, String tenantDomain) {
+
+        Map<String, String> claimProperties = getClaimProperties(tenantDomain, claimURI);
```
Comment (indeewari) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1146]: Is this null safe ? can claimProperties be null ?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java [hunk 3]
```diff
@@ -1129,4 +1130,27 @@ public static Map<String, String> getClaimProperties(String tenantDomain, String
...
+        try {
+            if (claimProperties.containsKey(IdentityMgtConstants.ENABLE_ENCRYPTION)) {
+                return claimValue;
+            }
+            return TOTPUtil.encrypt(claimValue);
```
Comment (indeewari) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1151]: Is this null and empty safe ?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 2 above)
Comment (lashinijay) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1146]: No claimProperties cannot be null. getClaimProperties will return an empty map if no properties are available. [https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/159/files#diff-69aacd3727454234e50c60bafac6f3fa57cf7154db56021ced438a640608cde1R1121](https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/159/files#diff-69aacd3727454234e50c60bafac6f3fa57cf7154db56021ced438a640608cde1R1121)

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java (hunk 3 above)
Comment (lashinijay) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1151]: Ideally, this cannot be null or empty, as this is a system-generated value. However, will add an empty check to be safe.

Comment (indeewari) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1154]: Can we use StringUtil isNotBlank ?

Comment (lashinijay) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtil.java:1154]: done

**Review by indeewari — APPROVED** [review]

**Review by janakamarasena — CHANGES_REQUESTED** [review]
Please include unit tests

**Comment by lashinijay** [comment]
> Please include unit tests

There is no functional changes introduced with the refactoring and there are already existing unit tests covering this functionality.

**Comment by indeewari** [comment]
The PR failure is due to uddi pom checksum failure while downloading from Nexus. This is being investigated by the digiops team now. Email: **Corrupted pom file in WSO2 Nexus**

**Comment by janakamarasena** [comment]
> 
> There are no functional changes introduced with the refactoring and there are existing unit tests covering this functionality.
> 
> https://github.com/wso2-extensions/identity-outbound-auth-totp/blob/master/component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPKeyGeneratorTest.java#L66
> 

Is this test covering the possible cases for the newly introduced method `getProcessedClaimValue`? Went through the code a bit and it doesn't seem to cover it.

**Comment by lashinijay** [comment]
> > There are no functional changes introduced with the refactoring and there are existing unit tests covering this functionality.
> > https://github.com/wso2-extensions/identity-outbound-auth-totp/blob/master/component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPKeyGeneratorTest.java#L66
> 
> Is this test covering the possible cases for the newly introduced method `getProcessedClaimValue`? Went through the code a bit and it doesn't seem to cover it.

The `getProcessedClaimValue()` this method will return a string (encrypted or plain text) and it is covered by `testGenerateClaims()` method by checking whether it is null or not.

File: component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java [hunk 4]
```diff
@@ -601,4 +605,48 @@ public IObjectFactory getObjectFactory() {
...
+                claimProperties_3);
+        localClaims3.add(localClaim3);
+
+        return new Object[][]{
+                {"http://wso2.org/claims/identity/secretkey", localClaims1, "AER2BRI0LK4XCSC1", "AER2BRI0LK4XCSC1"},
+                {"http://wso2.org/claims/identity/secretkey", localClaims1, "", ""},
+                {"http://wso2.org/claims/identity/secretkey", localClaims2, "AER2BRI0LK4XCSC1", "AER2BRI0LK4XCSC1"},
+                {"http://wso2.org/claims/identity/secretkey", localClaims2, "", ""},
+//                {"http://wso2.org/claims/identity/secretkey", localClaims3, "AER2BRI0LK4XCSC1", null},
+//                {"http://wso2.org/claims/identity/secretkey", localClaims3, "", ""},
```
Comment (dewniMW) [inline component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java:649]: why are these commented?

File: component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java [hunk 5]
```diff
@@ -601,4 +605,48 @@ public IObjectFactory getObjectFactory() {
...
+        assertEquals(TOTPUtil.getProcessedClaimValue(claimURI, claimValue,"testDomain"), expectedClaimValue);
+    }
+
+    @DataProvider(name = "processedClaimValueTestDataProvider")
+    public static Object[][] getProcessedClaimValueTestData() {
+
+        HashMap<String, String> claimProperties_1 = new HashMap<>();
+        claimProperties_1.put("EnableEncryption", "true");
+        List<LocalClaim> localClaims1 = new ArrayList<>();
+        LocalClaim localClaim1 = new LocalClaim("http://wso2.org/claims/identity/secretkey", null,
```
Comment (dewniMW) [inline component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java:626]: `http://wso2.org/claims/identity/secretkey` -> isn't there any constant defined for this?

**Comment by lashinijay** [comment]
Closing this PR as this will be covered with https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/162

File: component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java (hunk 5 above)
Comment (lashinijay) [reply component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java:626]: Will be addressed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/162

File: component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java (hunk 4 above)
Comment (lashinijay) [reply component/authenticator/src/test/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPUtilTest.java:649]: will be addressed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/162

**Pull request closed without merging**